### ISSUE TITLE
Mark internal primitives as `/** @psalm-immutable */`?

### ISSUE THREAD

**Issue by Ocramius** [Innmind/Immutable#6]
Looking at various primitives in this library, there is a focus on purity and immutability in general: it would probably be beneficial to mark all applicable classes as `/** @psalm-immutable */`.

**Comment by Ocramius** [comment]
An example of such addition can be found at https://github.com/ramsey/uuid/pull/281

**Comment by Baptouuuu** [comment]
Hi, I've added this annotation in #7  but I have an interrogation on this point: 
- `Set` and `Sequence` have a lazy method that use a function returning generators that will provide the elements of the structures. This is helpful to generate structures that can't fit in memory (such as an API paginated list), it acts like an immutable structure but typewise it is not since 2 iterations on the structure may generate different results. For this case I am not sure it's okay that the structure is declared immutable

I'd like your opinion on this.

**Comment by Ocramius** [comment]
If the two iterations are over semantically equal iterators, I don't see a problem?

Important in purity are:

 * referential transparency (same results over repeated calls, regardless of system-wide changes)
 * deterministic execution time

**Comment by Baptouuuu** [comment]
Take the following example: 

```php
Sequence::lazy(
    Line::class, // imaginary immutable class
    function() use ($someFileResource): \Generator {
        \fseek($someFileResource);
        while (!\feof($someFileResource)) {
            yield new Line(\fgets($someFileResource));
        }
    }
);
```
This snippet would return a `Sequence` of `Line`s, my interrogation is that in themselves `Sequence` and `Line` are immutable but two iterations on this sequence would yield 2 different objects for each line, and it even may different sequences as the file _may_ change between the 2 iterations.

But maybe I mix the fact that the `Sequence` class is immutable but the object isn't, I don't know.

The two points you listed tend to let me think it isn't.

**Comment by Ocramius** [comment]
yeah, that's tricky. ideally, we would have the implementation as pure/impure depending on wrapped consteuct purity 🤔

probably best to leave this out (for now) for any non-strict (lazy) structure

**Comment by Baptouuuu** [comment]
The implementations are different between _standard_ and lazy but the common interface is not exposed to the end user, only the final class wrapping the implementation is. 

I've changed this library a few months back to have this behaviour. Prior to that I had an interface for each structure so I could have different implementations exposed to the end user with only some annotated as immutable. But having this model prevented me to add new behaviours to the structures without introducing BC breaks. Since I use this library in about 70 other ones, the maintainability cost was too high.

As far as this library is currently built I'm afraid I can't really declare the structures truly immutable, at least not without lying to a certain degree to the end user. I'm not sure such a _lie_ is worth it, sadly.

**Comment by Ocramius** [comment]
Makes sense: let's close for now, and we can re-consider later, maybe with more powerful inference around immutability.

**Comment by Baptouuuu** [comment]
Little update on recent work to declare the structures immutable. 

Turns out it's the deferred structures that pose more problems for immutability as it contains state (keeping generated values in memory). But since the final behaviour for the user is to always have the same values I decided to suppress the psalm errors for this kind of structures.

For the lazy structures, immutability is not a problem as the immutable code is done outside the structures (see IO monads in functional languages).
